src/shared/sol-conffile.c

```diff
@@ -58,87 +60,207 @@ sol_conffile_entry_sort_cb(const void *data1, const void *data2)
 {
-    struct sol_ptr_vector pv;
-    gchar **groups, **groups_ptr;
-    struct sol_conffile_entry *entry = NULL;
+    free(entry->id);
+    free(entry->type);
+    if (entry->options) {
+        char **tmp = entry->options;
+        while (*tmp) {
+            free(*tmp);
+            tmp++;
+        }
+        free(entry->options);
+    }
+    free(entry);
+}
 
-    sol_ptr_vector_init(&pv);
+static int
+sol_conffile_set_entry_options(struct sol_conffile_entry *entry, struct sol_json_token options_object)
+{
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    struct sol_ptr_vector vec_options;
+    enum sol_json_loop_reason reason;
+
+    sol_ptr_vector_init(&vec_options);
+
+    sol_json_scanner_init_from_token(&scanner, &options_object);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&scanner, &token, &key, &value, reason) {
+        int key_len, value_len;
+        char *tmp;
+        key_len = sol_json_token_get_size(&key) - 2; // remove quotes from the key.
+        value_len = sol_json_token_get_size(&value); // do not remove quotes from the value. (not everything is a string here)
+        if (asprintf(&tmp, "%.*s=%.*s", key_len, key.start + 1, value_len, value.start) <= 0) {
+            SOL_WRN("Couldn't allocate memory for the config file, ignoring options.");
+            return -ENOMEM;
+            break;
+        }
+        sol_ptr_vector_append(&vec_options, tmp);
+    }
+    if (reason != SOL_JSON_LOOP_REASON_OK) {
+        int i;
+        void *ptr;
+        SOL_DBG("Error: Invalid Json.");
+        SOL_PTR_VECTOR_FOREACH_IDX(&vec_options, ptr, i) {
+            free(ptr);
+        }
+        sol_ptr_vector_clear(&vec_options);
+        sol_free_conffile_entry(entry);
+        return -ENOKEY;
+    }
 
-    groups = g_key_file_get_groups(keyfile, NULL);
-    for (groups_ptr = groups; *groups_ptr; groups_ptr++) {
-        entry = calloc(1, sizeof(*entry));
-        SOL_NULL_CHECK(entry, pv);
+    sol_ptr_vector_append(&vec_options, NULL);
+    entry->options = vec_options.base.data;
+    return 0;
+}
 
-        entry->id = strdup(*groups_ptr);
+static char *
+sol_dup_json_str(struct sol_json_token token)
+{
+    char *ret = sol_json_token_get_size(&token) >= 2 ?
+        strndup(token.start + 1, sol_json_token_get_size(&token) - 2) : NULL;
 
-        entry->type = g_key_file_get_string(keyfile,
-            *groups_ptr, "Type", NULL);
+    if (!ret)
+        SOL_WRN("Error alocating memory for string");
+    return ret;
+}
 
-        if (!entry->type) {
-            SOL_DBG("could not find mandatory 'Type' key on group [%s],"
-                " skipping entry", *groups_ptr);
-            free((void *)entry->id);
-            free(entry);
-            continue;
+static int
+sol_conffile_json_to_vector(struct sol_json_scanner scanner, struct sol_ptr_vector *pv)
+{
+    struct sol_conffile_entry *entry = NULL;
+    const char *node_group = "nodetypes";
```

Comment:
`static const char *node_group = "..."`: this way, the array is always kept in the text segment.